src/main/java/de/hysky/skyblocker/skyblock/dungeon/puzzle/waterboard/WaterboardOneFlow.java

```diff
@@ -0,0 +1,555 @@
+			if (leverType != null) {
+				firstSwitches.add(leverType);
+			}
+		}
+
+		if (firstSwitches.contains(LeverType.GOLD) && firstSwitches.contains(LeverType.TERRACOTTA)) {
+			return 1;
+		} else if (firstSwitches.contains(LeverType.EMERALD) && firstSwitches.contains(LeverType.QUARTZ)) {
+			return 2;
+		} else if (firstSwitches.contains(LeverType.QUARTZ) && firstSwitches.contains(LeverType.DIAMOND)) {
+			return 3;
+		} else if (firstSwitches.contains(LeverType.GOLD) && firstSwitches.contains(LeverType.QUARTZ)) {
+			return 4;
+		}
+
+		return 0;
+	}
+
+	private String findDoors() {
+		// Determine which doors are closed
+		StringBuilder doorBuilder = new StringBuilder();
+		BlockPos.Mutable doorPos = new BlockPos.Mutable(15, 57, 19);
+		for (int i = 0; i < 5; i++) {
+			if (!world.getBlockState(room.relativeToActual(doorPos)).isAir()) {
+				doorBuilder.append(i);
+			}
+			doorPos.move(Direction.NORTH);
+		}
+		return doorBuilder.toString();
+	}
+
+	private boolean checkWater() {
+		// Make sure there is no water currently in the board
+		for (int x = BOARD_MIN_X; x <= BOARD_MAX_X; x++) {
+			for (int y = BOARD_MIN_Y; y <= BOARD_MAX_Y; y++) {
+				BlockPos pos = room.relativeToActual(new BlockPos(x, y, BOARD_Z));
+				BlockState state = world.getBlockState(pos);
+				if (state.isOf(Blocks.WATER)) {
+					return false;
+				}
+			}
+		}
+		return true;
+	}
+
+	private Map<LeverType, List<Double>> setupSolution(JsonObject data) {
+		Map<LeverType, List<Double>> solution = makeEmptySolution();
+		for (Map.Entry<String, JsonElement> entry : data.entrySet()) {
+			LeverType leverType = LeverType.fromName(entry.getKey());
+			if (leverType != null) {
+				List<Double> times = new ArrayList<>();
+				for (JsonElement element : entry.getValue().getAsJsonArray()) {
+					times.add(element.getAsDouble());
+				}
+				solution.put(leverType, times);
+			}
+		}
+
+		// If the solver was reset after using some levers, make sure they are flipped back to the correct positions
+		for (LeverType leverType : LeverType.values()) {
+			List<Double> times = solution.get(leverType);
+			if (leverType != LeverType.WATER && isLeverActive(leverType)) {
+				if (times.isEmpty() || times.getFirst() != 0.0) {
+					times.addFirst(0.0);
+				} else {
+					times.removeFirst();
+				}
+			}
+		}
+
+		return solution;
+	}
+
+	private boolean isLeverActive(LeverType leverType) {
+		BlockPos offset = leverType.initialPositions[variant - 1];
+		if (offset == null) {
+			return false;
+		}
+		return !world.getBlockState(room.relativeToActual(WATER_ENTRANCE_POSITION.add(offset))).isOf(leverType.block);
+	}
+
+	private boolean isPuzzleSolved() {
+		if (doors == null || initialDoors == null || waterStartMillis == 0) {
+			return false;
+		}
+		String currentDoors = findDoors();
+		for (int i = 0; i < 5; i++) {
+			String s = String.valueOf(i);
+			if (initialDoors.contains(s) == currentDoors.contains(s) == doors.contains(s)) {
```

Comment:
This is a bit confusing, but I think it makes sense. Do you have a way of explaining this? Can you comment this?

Reply:
Try to imagine a truth table for all the cases and it should make sense. I will add a comment to describe it in words.